src/main/java/org/embulk/filter/column/JsonColumn.java

```diff
@@ -39,33 +44,65 @@ public JsonColumn(String path, Type type, Value defaultValue)
         this.type = type;
         this.defaultValue = (defaultValue == null ? ValueFactory.newNil() : defaultValue);
-        this.src = (src == null ? path : src);
+        this.src = compiledSrc.toString();
 
         this.pathValue = ValueFactory.newString(path);
-        this.parentPath = parentPath(path);
-        this.baseName = baseName(path);
-        if (this.baseName.equals("[*]")) {
+        this.parentPath = compiledPath.getParentPath();
+
+        if (compiledRoot.getTailPath().equals("[*]")) {
             throw new ConfigException(String.format("%s wrongly ends with [*], perhaps you can remove the [*]", path));
         }
-        this.baseIndex = baseIndex(path);
+        this.tailIndex = tailIndex(compiledRoot);
         this.parentPathValue = ValueFactory.newString(parentPath);
-        this.baseNameValue = ValueFactory.newString(baseName);
+        String tailName = getTailName(compiledRoot);
+        this.tailNameValue = tailName == null ? ValueFactory.newNil() : ValueFactory.newString(tailName);
 
         this.srcValue = ValueFactory.newString(this.src);
-        this.srcParentPath = parentPath(this.src);
-        this.srcBaseName = baseName(this.src);
-        this.srcBaseIndex = baseIndex(this.src);
+        this.srcParentPath = compiledSrc.getParentPath();
+        this.srcTailIndex = tailIndex(compiledSrcRoot);
         this.srcParentPathValue = ValueFactory.newString(this.srcParentPath);
-        this.srcBaseNameValue = ValueFactory.newString(this.srcBaseName);
+        String srcTailName = getTailName(compiledSrcRoot);
+        this.srcTailNameValue = srcTailName == null ? ValueFactory.newNil() : ValueFactory.newString(srcTailName);
 
-        if (! srcParentPath.equals(parentPath)) {
+        if (!srcParentPath.equals(parentPath)) {
             throw new ConfigException(String.format("The branch (parent path) of src \"%s\" must be same with of name \"%s\" yet", src, path));
         }
     }
 
+    // $['foo'] or $.foo => foo
+    // $['foo'][0] or $.foo[0] or $['foo'][*] or $.foo[*] => null
+    private String getTailName(RootPathToken root)
+    {
+        PathToken pathToken = root.getTail();
+        if (pathToken instanceof PropertyPathToken) {
+            if (!((PropertyPathToken) pathToken).singlePropertyCase()) {
+                throw new ConfigException(String.format("Multiple property is not supported \"%s\"", root.toString()));
+            }
+            return ((PropertyPathToken) pathToken).getProperties().get(0);
+        } else {
+            return null;
+        }
+    }
+
+    private Long tailIndex(RootPathToken root)
+    {
+        PathToken tail = root.getTail();
+        if (tail instanceof ArrayPathToken) {
+            ArrayIndexOperation arrayIndexOperation = ((ArrayPathToken) tail).getArrayIndexOperation();
+            PathTokenUtil.assertSupportedArrayPathToken(arrayIndexOperation, path);
+            return arrayIndexOperation.indexes().get(0).longValue();
+        } else {
```

Comment:
Embulk uses code style [Airlift](https://github.com/airlift/codestyle), and its else statement must looks like

```
}
else {
```